src/components/compress/index.tsx

```diff
@@ -0,0 +1,408 @@
+        type: newType,
+        options: encoderMap[newType].defaultOptions,
+      }),
+    });
+  }
+
+  private onPreprocessorOptionsChange(index: 0 | 1, options: PreprocessorState): void {
+    this.setState({
+      images: cleanSet(this.state.images, `${index}.preprocessorState`, options),
+    });
+  }
+
+  private onEncoderOptionsChange(index: 0 | 1, options: EncoderOptions): void {
+    this.setState({
+      images: cleanSet(this.state.images, `${index}.encoderState.options`, options),
+    });
+  }
+
+  componentWillReceiveProps(nextProps: Props): void {
+    if (nextProps.file !== this.props.file) {
+      this.updateFile(nextProps.file);
+    }
+  }
+
+  componentDidUpdate(prevProps: Props, prevState: State): void {
+    const { source, images } = this.state;
+
+    for (const [i, image] of images.entries()) {
+      const prevImage = prevState.images[i];
+      const sourceChanged = source !== prevState.source;
+      const encoderChanged = image.encoderState !== prevImage.encoderState;
+      const preprocessorChanged = image.preprocessorState !== prevImage.preprocessorState;
+
+      // The image only needs updated if the encoder/preprocessor settings have changed, or the
+      // source has changed.
+      if (sourceChanged || encoderChanged || preprocessorChanged) {
+        if (prevImage.downloadUrl) URL.revokeObjectURL(prevImage.downloadUrl);
+        this.updateImage(i, {
+          skipPreprocessing: !sourceChanged && !preprocessorChanged,
+        }).catch((err) => {
+          console.error(err);
+        });
+      }
+    }
+  }
+
+  private onCopyToOtherClick(index: 0 | 1) {
+    const otherIndex = (index + 1) % 2;
+
+    this.setState({
+      images: cleanSet(this.state.images, otherIndex, this.state.images[index]),
+    });
+  }
+
+  @bind
+  private async updateFile(file: File | Fileish) {
+    this.setState({ loading: true });
+
+    try {
+      let data: ImageData;
+      let vectorImage: HTMLImageElement | undefined;
+
+      // Special-case SVG. We need to avoid createImageBitmap because of
+      // https://bugs.chromium.org/p/chromium/issues/detail?id=606319.
+      // Also, we cache the HTMLImageElement so we can perform vector resizing later.
+      if (file.type.startsWith('image/svg+xml')) {
+        vectorImage = await processSvg(file);
+        data = drawableToImageData(vectorImage);
+      } else {
+        data = await decodeImage(file);
+      }
+
+      let newState: State = {
+        ...this.state,
+        source: { data, file, vectorImage },
+        loading: false,
+      };
+
+      for (const i of [0, 1]) {
+        // Ditch previous encodings
+        const downloadUrl = this.state.images[i].downloadUrl;
+        if (downloadUrl) URL.revokeObjectURL(downloadUrl!);
+
+        newState = cleanMerge(newState, `images.${i}`, {
+          preprocessed: undefined,
+          file: undefined,
+          downloadUrl: undefined,
+          data: undefined,
+        });
```

Comment:
This fixes the issue where you'd see two different halves of an image when loading a new image.